### PR TITLE
merge windows and linux build scripts, simplify electron build steps

### PR DESCRIPTION
**Pull request by trevj**
Main thing here is to share build scripts for the Windows and Linux clients; while I was working on it, I was also able to simplify the overall build process.

The key changes are:
- specify electron-builder options in `package.json`, where possible
- have electron-builder build from the "top" of the repo rather than `electron/build`

This let me share most options between Windows and Linux and eliminated a **lot** of manual file copying, **especially** `www/`.

Still a few rough edges but I think it's a huge improvement overall.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -65,5 +65,6 @@
   },
   "resolutions": {
     "natives": "1.1.6"
-  }
-}
+  },
+  "main": "build/electron/electron/index.js"
```
Comment (fortuna) [inline package.json]: I find it odd to have this at the root of the repo. Can we put in the electron directory or pass as a build command line flag?

File: src/electron/build_action.sh [hunk 2]
```diff
@@ -17,48 +17,7 @@
...
-# Not strictly necessary when running from the command line but this sets fields
-# such as productName which influences things like the appData directory, easing
-# debugging.
-cp package.json $OUTPUT
-
-# Icons.
-electron-icon-maker --input=src/electron/logo.png --output=build/windows
+# Environment variables.
+# TODO: make non-packaged builds work without this
+scripts/environment_json.sh -p electron > www/environment.json
```
Comment (fortuna) [inline src/electron/build_action.sh]: Where's the actual build call?

File: src/electron/package_common_action.sh [hunk 3]
```diff
@@ -14,18 +14,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-yarn do src/electron/linux/build
+yarn do src/electron/build
```
Comment (fortuna) [inline src/electron/package_common_action.sh]: Can't we move everything here to src/electron/build? Why do we need `package_common_action.sh`?

File: src/electron/process_manager.ts [hunk 4]
```diff
@@ -240,7 +242,7 @@ function startLocalShadowsocksProxy(
     // Note that if you run with -v then ss-local may output a lot of data to stderr which
     // will cause the binary to fail:
     //   https://nodejs.org/dist/latest-v10.x/docs/api/child_process.html#child_process_maxbuffer_and_unicode
-    ssLocal = spawn(pathToEmbeddedBinary('ss-local'), ssLocalArgs);
+    ssLocal = spawn(pathToEmbeddedBinary('shadowsocks-libev', 'ss-local'), ssLocalArgs);
```
Comment (fortuna) [inline src/electron/process_manager.ts]: Did ss-local get renamed? Isn't it clearer to keep the original name?

File: src/electron/build_action.sh (hunk 2 above)
Comment (trevj) [reply src/electron/build_action.sh]: `tsc`, above.

File: src/electron/package_common_action.sh (hunk 3 above)
Comment (trevj) [reply src/electron/package_common_action.sh]: Oh, for faster builds. The icon step, in particular, takes several seconds.

File: src/electron/process_manager.ts (hunk 4 above)
Comment (trevj) [reply src/electron/process_manager.ts]: The layout on disk changed: previously, everything was jammed into the top of the install directory. Now, the directory tree is mirrored there - so the paths have to be more exact but do not change between development and install time.

File: package.json (hunk 1 above)
Comment (trevj) [reply package.json]: I don't know of a good alternative. `electron-builder` and `electron` have options to override the *root folder* but not the main *class* which has to be at a certain location within that root folder (we *could* add a one-liner at `/index.js` - we currently do that and I was happy to eliminate it in this PR).

I think switching to workspaces is the real solution here but I definitely don't want to tackle that right now.

**Comment by trevj** [comment]
@fortuna I don't know of an alternative to the layering with our current build system. Not sure I can think of a better than `package_common`. This is very close to what we have in the server manager, where it's been working pretty well.

File: .gitignore [hunk 5]
```diff
@@ -10,7 +10,6 @@ info.txt
 /third_party/Potatso/Pods
 /Carthage
 xcuserdata/
-dist/
```
[the remark this answers is not in the document]
Comment (fortuna) [reply .gitignore]: I think the way it is it ignores the directory and subdirectories? Anyway, I just wanted to check if there aren't dist/ directories somewhere else.

**Pull request merged**